internal/gps/pkgtree/digest_test.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pkgtree
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+// crossBuffer is a test io.Reader that emits a few canned responses.
+type crossBuffer struct {
+	readCount  int
+	iterations []string
+}
+
+func (cb *crossBuffer) Read(buf []byte) (int, error) {
+	if cb.readCount == len(cb.iterations) {
+		return 0, io.EOF
+	}
+	cb.readCount++
+	return copy(buf, cb.iterations[cb.readCount-1]), nil
+}
+
+func streamThruLineEndingReader(t *testing.T, iterations []string) []byte {
+	dst := new(bytes.Buffer)
+	n, err := io.Copy(dst, newLineEndingReader(&crossBuffer{iterations: iterations}))
+	if got, want := err, error(nil); got != want {
+		t.Errorf("(GOT): %v; (WNT): %v", got, want)
+	}
+	if got, want := n, int64(dst.Len()); got != want {
+		t.Errorf("(GOT): %v; (WNT): %v", got, want)
+	}
+	return dst.Bytes()
+}
+
+func TestLineEndingReader(t *testing.T) {
+	testCases := []struct {
+		input  []string
+		output string
+	}{
+		{[]string{"\r"}, "\r"},
+		{[]string{"\r\n"}, "\n"},
+		{[]string{"now is the time\r\n"}, "now is the time\n"},
+		{[]string{"now is the time\r\n(trailing data)"}, "now is the time\n(trailing data)"},
+		{[]string{"now is the time\n"}, "now is the time\n"},
+		{[]string{"now is the time\r"}, "now is the time\r"},     // trailing CR ought to convey
+		{[]string{"\rnow is the time"}, "\rnow is the time"},     // CR not followed by LF ought to convey
+		{[]string{"\rnow is the time\r"}, "\rnow is the time\r"}, // CR not followed by LF ought to convey
+
+		{[]string{"this is the result\r\nfrom the first read\r", "\nthis is the result\r\nfrom the second read\r"},
```

Comment:
so, this case checks for correct behavior when a CRLF breaks across reads. excellent. but, with the way the algorithm carries state from one read to the next, it'd be great to see a few more test cases, just to be sure:

1. three read segments, where 1->2 and 2->3 both break across a CRLF
2. three read segments, where 1->2 breaks across CRLF and 2->3 does not
3. three read segments, where 1->2 breaks across CRLF and 2 ends in CR but 3 does not begin LF
4. three read segments, where 1 ends in CR but 2 does not begin LF, and 2->3 breaks across CRLF
5. three read segments, where 1 ends in CR but 2 does not begin LF, and 2->3 does not break across CRLF
6. four read segments, where 1->2 and 2->3 both break across a CRLF, but 3->4 does not

i'm just trying to put some checks into place that guard us against the possibility of weird OBO errors arising from the end state of a previous read affecting the bookkeeping of a subsequent read. unless some other test cases are covering that sort of thing, and i missed it?